core/domain/stats_jobs_one_off.py

```diff
@@ -18,14 +18,403 @@
+                'submitted_answer_list': item.submitted_answer_list
+            }
+            yield(item.exploration_id, value)
+
+    @staticmethod
+    def reduce(key, stringified_values):
+        num_starts = 0
+        num_completions = 0
+
+        exp_id = key
+        exploration = exp_services.get_exploration_by_id(exp_id)
+        latest_exp_version = exploration.version
+
+        # Check if model already exists. If it does, delete it.
+        exploration_stats_model = stats_models.ExplorationStatsModel.get_model(
+            exp_id, latest_exp_version)
+        if exploration_stats_model is not None:
+            exploration_stats_model.delete()
+
+        # The list of state hit events in stringified_values.
+        values_state_hit = []
+        # The list of state answers instances in stringified_values.
+        values_state_answer = []
+        # The list of session_ids of learners completing an exploration.
+        completed_session_ids = []
+        for value_str in stringified_values:
+            value = ast.literal_eval(value_str)
+
+            if value['event_type'] == feconf.EVENT_TYPE_START_EXPLORATION:
+                num_starts += 1
+            elif value['event_type'] == feconf.EVENT_TYPE_COMPLETE_EXPLORATION:
+                num_completions += 1
+                completed_session_ids.append(value['session_id'])
+            elif value['event_type'] == feconf.EVENT_TYPE_STATE_HIT:
+                values_state_hit.append(value)
+            elif value['event_type'] == (
+                    MigrateStatistics.EVENT_TYPE_STATE_ANSWERS):
+                values_state_answer.append(value)
+
+        state_stats_mapping = {}
+        # Sort list of state hit events in increasing order of it's timestamp.
+        values_state_hit = sorted(values_state_hit, key=lambda k: k[
+            'created_on'])
+        for version in range(1, latest_exp_version+1):
+            values_state_hit_versioned = [
+                val for val in values_state_hit if val['version'] == version]
+            values_state_answer_versioned = [
+                val for val in values_state_answer if val['version'] == version]
+            exploration_versioned = exp_services.get_exploration_by_id(
+                exp_id, version)
+
+            if version == 1:
+                # Create default state stats mapping for the first version.
+                for state_name in exploration_versioned.states:
+                    state_stats_mapping[state_name] = (
+                        stats_domain.StateStats.create_default())
+            else:
+                exp_commit_log = exp_models.ExplorationCommitLogEntryModel.get(
+                    'exploration-%s-%s' % (exp_id, version))
+                change_list = exp_commit_log.commit_cmds
+
+                # Handling state additions, renames and deletions.
+                for change_dict in change_list:
+                    if change_dict['cmd'] == exp_domain.CMD_ADD_STATE:
```

Comment:
One thought: if I have states 1 --> 2 --> 3 --> 4 --> 5 and later on add a state 3a after 3, the state hit counts for 3a will be much lower than those for 4. Is this weird? Are the hit counts still meaningful in that case, from a user's perspective?

Reply:
Let's leave it as-is for now, I think. We can do something more sophisticated later.